File: api/internal/api/handlers/outbound_calls.go
package handlers

import (
	"claimsio/internal/config"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"sync"

	"github.com/gorilla/websocket"
	"github.com/twilio/twilio-go"

	// "github.com/twilio/twilio-go/twiml"
	twilioApi "github.com/twilio/twilio-go/rest/api/v2010"

	"go.uber.org/zap"
)

// TODO - test this

type OutboundConversation struct {
	CallSid        string
	ConversationID string
	Number         string
	UserData       map[string]interface{}
}

var outboundConversations sync.Map

func HandleOutboundCall(cfg *config.Config) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var req struct {
			Number string `json:"number"`
			Prompt string `json:"prompt"`
		}

		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			http.Error(w, "Invalid request body", http.StatusBadRequest)
			return
		}

		if req.Number == "" {
			http.Error(w, "Phone number is required", http.StatusBadRequest)
			return
		}

		// Create Twilio call
		call, err := createTwilioCall(req.Number, req.Prompt, r.Host, cfg.TwilioPhoneNumber)
		if err != nil {
			zap.L().Error("Failed to create Twilio call", zap.Error(err))
			http.Error(w, "Failed to initiate call", http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"success": true,
			"message": "Call initiated",
			"callSid": call.Sid,
		})
	})
}

func HandleOutboundCallTwiml(cfg *config.Config) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		prompt := r.URL.Query().Get("prompt")
		number := r.URL.Query().Get("number")

		twiml := fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
        <Response>
            <Connect>
                <Stream url="wss://%s/outbound-media-stream">
                    <Parameter name="prompt" value="%s" />
                    <Parameter name="number" value="%s" />
                </Stream>
            </Connect>
        </Response>`, r.Host, url.QueryEscape(prompt), url.QueryEscape(number))

		w.Header().Set("Content-Type", "text/xml")
		w.Write([]byte(twiml))
	})
}

// from docs - can be improved
// router.POST("/answer", func(context *gin.Context) {
// 	say := &twiml.VoiceSay{
// 		Message: "Hello from your pals at Twilio! Have fun.",
// 	}

// 	twimlResult, err := twiml.Voice([]twiml.Element{say})
// 	if err != nil {
// 		context.String(http.StatusInternalServerError, err.Error())
// 	} else {
// 		context.Header("Content-Type", "text/xml")
// 		context.String(http.StatusOK, twimlResult)
// 	}
// })

func HandleOutboundMediaStream(cfg *config.Config, upgrader websocket.Upgrader) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		conn, err := upgrader.Upgrade(w, r, nil)
		if err != nil {
			zap.L().Error("Failed to upgrade connection", zap.Error(err))
			return
		}
		defer conn.Close()

		zap.L().Info("New outbound WebSocket connection established")

		var streamSid string
		var callSid string
		var elevenLabsWs *websocket.Conn
		var customParameters map[string]interface{}
		isDisconnecting := false

		for {
			messageType, message, err := conn.ReadMessage()
			if err != nil {
				zap.L().Error("Error reading message", zap.Error(err))
				break
			}

			if messageType != websocket.TextMessage {
				continue
			}

			var data map[string]interface{}
			if err := json.Unmarshal(message, &data); err != nil {
				zap.L().Error("Error parsing message", zap.Error(err))
				continue
			}

			event, ok := data["event"].(string)
			if !ok {
				continue
			}

			switch event {
			case "start":
				startData := data["start"].(map[string]interface{})
				streamSid = startData["streamSid"].(string)
				callSid = startData["callSid"].(string)
				customParameters = startData["customParameters"].(map[string]interface{})

				// check user data
				userData, err := checkUserExists(customParameters["number"].(string))
				if err != nil {
					zap.L().Error("Failed to check user", zap.Error(err))
					return
				}

				// init ElevenLabs
				elevenLabsWs, err = initializeElevenLabs(customParameters,
					userData,
					cfg.ElevenLabsAgentID,
					cfg.ElevenLabsAPIKey,
					streamSid,
				)
				if err != nil {
					zap.L().Error("Failed to initialize ElevenLabs", zap.Error(err))
					return
				}

				// store conversation data
				conv := &OutboundConversation{
					CallSid: callSid,
					Number:  customParameters["number"].(string),
				}
				outboundConversations.Store(callSid, conv)

			case "media":
				if elevenLabsWs != nil && !isDisconnecting {
					mediaData := data["media"].(map[string]interface{})
					payload := mediaData["payload"].(string)

					msg := map[string]interface{}{
						"user_audio_chunk": payload,
					}
					if err := elevenLabsWs.WriteJSON(msg); err != nil {
						zap.L().Error("Failed to send audio to ElevenLabs", zap.Error(err))
					}
				}

			case "stop":
				isDisconnecting = true
				if elevenLabsWs != nil {
					elevenLabsWs.WriteJSON(map[string]string{"type": "end_conversation"})
					elevenLabsWs.Close()
				}

				// Send final webhook
				if conv, ok := outboundConversations.Load(callSid); ok {
					payload := map[string]interface{}{
						"conversation_id": conv.(*OutboundConversation).ConversationID,
						"phone_number":    conv.(*OutboundConversation).Number,
						"call_sid":        conv.(*OutboundConversation).CallSid,
					}

					sendWebhook("outbound-calls", payload, cfg.N8NAuthToken)

					outboundConversations.Delete(callSid)
				}

				// Send disconnect signals
				conn.WriteJSON(map[string]interface{}{
					"event":     "mark_done",
					"streamSid": streamSid,
				})
				conn.WriteJSON(map[string]interface{}{
					"event":     "clear",
					"streamSid": streamSid,
				})
				return
			}
		}
	})
}

// private

func createTwilioCall(number, prompt, host, twilioPhoneNumber string) (*twilioApi.ApiV2010Call, error) {
	callURL := fmt.Sprintf("https://%s/outbound-call-twiml?prompt=%s&number=%s",
		host, url.QueryEscape(prompt), url.QueryEscape(number))

	client := twilio.NewRestClient()

	params := &twilioApi.CreateCallParams{}
	params.SetTo(number)
	params.SetFrom(twilioPhoneNumber)
	params.SetUrl(callURL)

	call, err := client.Api.CreateCall(params)
	if err != nil {
		return nil, fmt.Errorf("failed to create call: %v", err)
	}

	return call, nil
}

File: calls/outbound-calls.js
import WebSocket from "ws";
import Twilio from "twilio";

export function registerOutboundRoutes(fastify) {
  const {
    ELEVENLABS_API_KEY,
    ELEVENLABS_AGENT_ID,
    TWILIO_ACCOUNT_SID,
    TWILIO_AUTH_TOKEN,
    TWILIO_PHONE_NUMBER,
    N8N_AUTH_TOKEN,
  } = process.env;

  if (
    !ELEVENLABS_API_KEY ||
    !ELEVENLABS_AGENT_ID ||
    !TWILIO_ACCOUNT_SID ||
    !TWILIO_AUTH_TOKEN ||
    !TWILIO_PHONE_NUMBER ||
    !N8N_AUTH_TOKEN
  ) {
    throw new Error("Missing required environment variables");
  }

  const twilioClient = new Twilio(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN);
  const transcripts = new Map();

  class TranscriptManager {
    constructor(callSid) {
      this.callSid = callSid;
      this.conversationId = null;
      this.userData = null;
    }
  }

  async function checkUserData(number) {
    try {
      const response = await fetch(
        "https://n8n.claimsio.com/webhook/check_if_user_exist",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({ phone: number }),
        },
      );
      if (response.status === 200) {
        const data = await response.json();
        return { authorized: true, userData: data };
      }
      return { authorized: false, userData: null };
    } catch (error) {
      console.error("[Auth] Error checking user:", error);
      return { authorized: false, userData: null };
    }
  }

  async function getSignedUrl() {
    try {
      const response = await fetch(
        `https://api.elevenlabs.io/v1/convai/conversation/get_signed_url?agent_id=${ELEVENLABS_AGENT_ID}`,
        {
          method: "GET",
          headers: {
            "xi-api-key": ELEVENLABS_API_KEY,
          },
        },
      );

      if (!response.ok) {
        throw new Error(`Failed to get signed URL: ${response.statusText}`);
      }

      const data = await response.json();
      return data.signed_url;
    } catch (error) {
      console.error("Error getting signed URL:", error);
      throw error;
    }
  }

  async function sendWebhook(payload) {
    try {
      console.log("[Webhook] Sending webhook with token:", N8N_AUTH_TOKEN);
      const response = await fetch(
        "https://n8n.claimsio.com/webhook/outbound-calls",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: N8N_AUTH_TOKEN, // Removed Bearer prefix
          },
          body: JSON.stringify(payload),
        },
      );
      console.log("[Webhook] Response status:", response.status);
      if (!response.ok) {
        const errorText = await response.text();
        console.error("[Webhook] Error response:", errorText);
        throw new Error(`Webhook failed: ${errorText}`);
      }
    } catch (error) {
      console.error("[Webhook] Error:", error);
      throw error; // Re-throw to handle it in the calling function
    }
  }

  fastify.post("/outbound-call", async (request, reply) => {
    const { number, prompt } = request.body;

    if (!number) {
      return reply.code(400).send({ error: "Phone number is required" });
    }

    try {
      const call = await twilioClient.calls.create({
        from: TWILIO_PHONE_NUMBER,
        to: number,
        url: `https://${request.headers.host}/outbound-call-twiml?prompt=${encodeURIComponent(
          prompt,
        )}&number=${encodeURIComponent(number)}`,
      });

      reply.send({
        success: true,
        message: "Call initiated",
        callSid: call.sid,
      });
    } catch (error) {
      console.error("Error initiating outbound call:", error);
      reply.code(500).send({
        success: false,
        error: "Failed to initiate call",
      });
    }
  });

  fastify.all("/outbound-call-twiml", async (request, reply) => {
    const prompt = request.query.prompt || "";
    const number = request.query.number || "";

    const twimlResponse = `<?xml version="1.0" encoding="UTF-8"?>
      <Response>
        <Connect>
          <Stream url="wss://${request.headers.host}/outbound-media-stream">
            <Parameter name="prompt" value="${prompt}" />
            <Parameter name="number" value="${number}" />
          </Stream>
        </Connect>
      </Response>`;

    reply.type("text/xml").send(twimlResponse);
  });

  fastify.register(async (fastifyInstance) => {
    fastifyInstance.get(
      "/outbound-media-stream",
      { websocket: true },
      (ws, req) => {
        console.info("[Server] Twilio connected to outbound media stream");

        let streamSid = null;
        let callSid = null;
        let elevenLabsWs = null;
        let customParameters = null;
        let transcriptManager = null;
        let isDisconnecting = false;

        async function disconnectTwilioCall() {
          if (isDisconnecting) return;

          isDisconnecting = true;
          console.log("[Twilio] Initiating call disconnect");

          if (transcriptManager?.conversationId) {
            try {
              const payload = {
                conversation_id: transcriptManager.conversationId,
                phone_number: customParameters?.number || "unknown",
                call_sid: callSid || "unknown",
              };
              console.log("[Webhook] Preparing to send payload:", payload);
              await sendWebhook(payload);
              console.log("[Webhook] Successfully sent webhook");
            } catch (error) {
              console.error("[Webhook] Failed to send webhook:", error);
              // Continue with disconnect even if webhook fails
            }
          }

          ws.send(
            JSON.stringify({
              event: "mark_done",
              streamSid,
            }),
          );

          ws.send(
            JSON.stringify({
              event: "clear",
              streamSid,
            }),
          );

          ws.send(
            JSON.stringify({
              event: "twiml",
              streamSid,
              twiml: "<Response><Hangup/></Response>",
            }),
          );

          setTimeout(() => {
            if (ws.readyState === WebSocket.OPEN) {
              ws.close();
            }
          }, 1000);
        }

        async function setupElevenLabsWithUserData(userData) {
          try {
            const signedUrl = await getSignedUrl();
            elevenLabsWs = new WebSocket(signedUrl);

            elevenLabsWs.on("open", () => {
              console.log("[ElevenLabs] Connected to Conversational AI");

              const initialConfig = {
                type: "conversation_initiation_client_data",
                conversation_config_override: {
                  agent: {
                    prompt: {
                      prompt: userData
                        ? `You are an outbound call agent. Context about the person you're calling:
Name: ${userData.debtor.first_name} ${userData.debtor.last_name}
Case Number: ${userData.case.case_number}
Debt Amount: ${userData.case.debt_amount} ${userData.case.currency}
Caller Phone: ${customParameters?.number},
Case Description: ${userData.case.case_description}

${customParameters?.prompt || ""}`
                        : customParameters?.prompt ||
                          "You are a customer service representative",
                    },
                    first_message: userData
                      ? `Hi ${userData.debtor.first_name}! I'm calling about your ${userData.case.case_number}. Do you have a moment to talk?`
                      : "Hello, do you have a moment to talk?",
                  },
                },
              };

              if (userData) {
                initialConfig.client_data = {
                  dynamic_variables: {
                    caller_phone: customParameters?.number || "unknown",
                    caller_name: `${userData.debtor.first_name} ${userData.debtor.last_name}`,
                    case_number: userData.case.case_number,
                    debt_amount: `${userData.case.debt_amount} ${userData.case.currency}`,
                    case_description: userData.case.case_description,
                  },
                };
              }

              console.log(
                "[ElevenLabs] Sending initial config:",
                JSON.stringify(initialConfig, null, 2),
              );
              elevenLabsWs.send(JSON.stringify(initialConfig));
            });

            elevenLabsWs.on("message", (data) => {
              try {
                const message = JSON.parse(data);

                switch (message.type) {
                  case "conversation_initiation_metadata":
                    if (
                      message.conversation_initiation_metadata_event
                        ?.conversation_id
                    ) {
                      transcriptManager.conversationId =
                        message.conversation_initiation_metadata_event.conversation_id;
                      console.log(
                        "[ElevenLabs] Stored conversation_id:",
                        transcriptManager.conversationId,
                      );
                    }
                    break;

                  case "audio":
                    if (streamSid && !isDisconnecting) {
                      const audioData = {
                        event: "media",
                        streamSid,
                        media: {
                          payload:
                            message.audio?.chunk ||
                            message.audio_event?.audio_base_64,
                        },
                      };
                      ws.send(JSON.stringify(audioData));
                    }
                    break;

                  case "interruption":
                    if (streamSid) {
                      ws.send(JSON.stringify({ event: "clear", streamSid }));
                    }
                    break;

                  case "ping":
                    if (message.ping_event?.event_id) {
                      elevenLabsWs.send(
                        JSON.stringify({
                          type: "pong",
                          event_id: message.ping_event.event_id,
                        }),
                      );
                    }
                    break;

                  case "end_of_conversation":
                    console.log("[ElevenLabs] End of conversation received");
                    disconnectTwilioCall();
                    break;

                  default:
                    console.log(
                      `[ElevenLabs] Unhandled message type: ${message.type}`,
                    );
                }
              } catch (error) {
                console.error("[ElevenLabs] Error processing message:", error);
              }
            });

            elevenLabsWs.on("error", (error) => {
              console.error("[ElevenLabs] WebSocket error:", error);
            });

            elevenLabsWs.on("close", (code, reason) => {
              console.log("[ElevenLabs] Disconnected with code:", code);
              console.log(
                "[ElevenLabs] Disconnect reason:",
                reason?.toString() || "No reason provided",
              );

              if (code === 1000 || code === 1005 || code === 1006) {
                console.log(
                  "[ElevenLabs] Normal closure detected, ending Twilio call",
                );
                disconnectTwilioCall();
              } else {
                console.log(
                  `[ElevenLabs] Abnormal closure (code: ${code}), handling gracefully`,
                );
              }
            });
          } catch (error) {
            console.error("[ElevenLabs] Setup error:", error);
          }
        }

        ws.on("error", console.error);

        ws.on("message", async (message) => {
          try {
            const msg = JSON.parse(message);
            console.log(`[Twilio] Received event: ${msg.event}`);

            if (isDisconnecting && msg.event !== "stop") {
              console.log(
                "[Twilio] Ignoring event during disconnect:",
                msg.event,
              );
              return;
            }

            switch (msg.event) {
              case "start":
                streamSid = msg.start.streamSid;
                callSid = msg.start.callSid;
                customParameters = msg.start.customParameters;
                transcriptManager = new TranscriptManager(callSid);
                transcripts.set(callSid, transcriptManager);

                console.log(
                  `[Twilio] Stream started - StreamSid: ${streamSid}, CallSid: ${callSid}`,
                );
                console.log("[Twilio] Start parameters:", customParameters);

                // Fetch user data before setting up ElevenLabs
                const { userData } = await checkUserData(
                  customParameters?.number,
                );
                transcriptManager.userData = userData;

                // Setup ElevenLabs with user data
                await setupElevenLabsWithUserData(userData);
                break;

              case "media":
                if (
                  elevenLabsWs?.readyState === WebSocket.OPEN &&
                  !isDisconnecting
                ) {
                  const audioMessage = {
                    user_audio_chunk: Buffer.from(
                      msg.media.payload,
                      "base64",
                    ).toString("base64"),
                  };
                  elevenLabsWs.send(JSON.stringify(audioMessage));
                }
                break;

              case "stop":
                if (elevenLabsWs?.readyState === WebSocket.OPEN) {
                  elevenLabsWs.send(
                    JSON.stringify({ type: "end_conversation" }),
                  );
                  elevenLabsWs.close();
                }
                await disconnectTwilioCall();
                break;

              default:
                console.log(`[Twilio] Unhandled event: ${msg.event}`);
            }
          } catch (error) {
            console.error("[Twilio] Error processing message:", error);
          }
        });

        ws.on("close", () => {
          console.log("[Twilio] Client disconnected");
          if (elevenLabsWs?.readyState === WebSocket.OPEN) {
            elevenLabsWs.close();
          }
        });
      },
    );
  });

  fastify.get("/transcript/:callSid", async (request, reply) => {
    const { callSid } = request.params;
    const transcript = transcripts.get(callSid);

    if (!transcript) {
      return reply.code(404).send({ error: "Transcript not found" });
    }

    reply.send(transcript);
  });
}
